File: Python/ByTier/D/1877_D_Effects_of_Anti_Pimples.py
# Problem: CF 1877 D - Effects of Anti Pimples
# https://codeforces.com/contest/1877/problem/D

"""
Code Purpose:
This code computes the sum of maximum values over all possible subsets of indices
chosen to be "black", where:
- Initially, all elements are white.
- Black elements are chosen from indices.
- Green elements are all white elements whose indices are multiples of at least one black index.
- Score is the maximum value among all black and green elements.
- We sum up scores for all 2^n - 1 possible choices of black indices.

Algorithm:
- For each index i, we compute the maximum value in the range [i, n] with step i (i.e., multiples of i).
- This is done by iterating from 1 to n and updating a[i] with max of a[i] and a[j] where j is a multiple of i.
- Then, we sort these values and compute the result using a weighted sum with powers of 2 (since each number can either be chosen or not in a subset).

Time Complexity: O(n log n) 
Space Complexity: O(n)

Techniques:
- Preprocessing to compute max values for multiples using sieve-like approach.
- Sorting and weighted sum using powers of 2.
"""

n = int(input())
a = [0] + [*map(int, input().split())]  # Add dummy 0 at start for 1-based indexing
mod = 998244353

# For each index i, propagate the max value to all its multiples
# This ensures a[i] holds the maximum value among all indices that are multiples of i
for i in range(1, n + 1):
    for j in range(i, n + 1, i):  # Iterate through all multiples of i
        a[i] = max(a[i], a[j])   # Update a[i] to be max between current and multiple

ans = 0
cnt = 1  # Count of subsets that include current element (starts with 1)

# Sort the processed values in ascending order for correct contribution calculation
for x in sorted(a[1:]):  # Iterate over elements from index 1 onwards
    ans = (ans + x * cnt) % mod  # Add contribution of current element
    cnt = (cnt * 2) % mod        # Double the count for next iteration (2^k subsets)

print(ans)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/D/1877_D_Effects_of_Anti_Pimples.cpp
// Problem: CF 1877 D - Effects of Anti Pimples
// https://codeforces.com/contest/1877/problem/D

/*
 * Problem: Effects of Anti Pimples
 * 
 * Purpose: Calculate the sum of scores for all possible ways to choose black indices,
 *          where each way results in a set of black and green elements, and score is
 *          the maximum value among these elements.
 * 
 * Algorithm:
 * - For each index i, we compute the maximum value among all elements at indices that are multiples of i.
 * - This helps determine the value that would be "green" if index i is chosen as black.
 * - Then, we count how many indices have the same maximum value (in c array).
 * - Using dynamic programming, we calculate powers of 2 to represent combinations.
 * - Finally, we compute the total contribution of each value to the answer by multiplying
 *   the number of subsets that include at least one index with that value, and summing over all values.
 * 
 * Time Complexity: O(n log n)
 * Space Complexity: O(n)
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
using ll = long long;
const int mod = 998244353;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(0);
  
  int n;
  cin >> n;
  vector<int> a(n);
  for (int &x : a) {
    cin >> x;
  }
  
  // c[i] will store how many indices have maximum value equal to i
  vector<int> c(100100);
  
  // Process each index to compute the maximum value accessible from it
  for (int i = 1; i <= n; i++) {
    for (int j = i; j <= n; j += i) {
      a[i - 1] = max(a[i - 1], a[j - 1]);  // Update maximum value at index i
    }
    c[a[i - 1]]++;  // Count how many indices have this max value
  }
  
  // Precompute powers of 2 modulo 998244353
  vector<int> p2(100100);
  p2[0] = 1;
  for (int i = 1; i <= n; i++) {
    p2[i] = (ll)p2[i - 1] * 2 % mod;
  }

  int tot = 0;  // Total count of indices processed so far
  int ans = 0;  // Final answer to be computed

  // For each possible value from 0 to 100000:
  for (int i = 0; i <= 100000; i++) {
    // Contribution = p2[tot] * (p2[c[i]] - 1) * i
    // p2[tot] = number of ways to choose black indices from indices not yet handled
    // (p2[c[i]] - 1) = number of non-empty subsets of indices with value i that can be chosen
    // i = value itself
    ans = (ans + (ll)p2[tot] * (p2[c[i]] - 1) % mod * i) % mod;
    tot += c[i];  // Increment total of processed indices
  }
  
  cout << ans;
}


// https://github.com/VaHiX/CodeForces/